[MRG] New Feature: VotingRegressor

stsouko opened the pull request:
#### Reference Issues/PRs

Fixes #10743

#### What does this implement/fix? Explain your changes.
Following @amueller suggestion about adding sklearn.ensemble.AverageRegressor, this PR implements:

- [x] sklearn.ensemble.AverageRegressor

- [x] Associated units tests

- [x] Docstrings

- [x] Examples

- [x] Updates the documentation

- [x] Updates what's new

#### Any other comments?
This PR is inspired from work on `VotingClassifier` and their associated unit-tests.

this is fork of PR #10868 which is not being finalized for a long time.

Commit:
AverageRegressor based on VotingClassifier implemented.
tests done.

this is fork of PR #10868 which is not being finalized for a long time.

Commit:
travis tests fixed.

Commit:
tutorial added.

Commit:
doctests fixed.
flake fixed.

eamanu commented:
Please add the [WIP] tag

Commit:
watsnew added.

Commit:
Merge remote-tracking branch 'sklearn/master'

# Conflicts:
#	doc/whats_new/v0.21.rst

Commit:
doc fixed.

eamanu reviewed (approved):
LGTM. I read the codes and IMO its ok.

jnothman commented on doc/modules/ensemble.rst:
```diff
@@ -1084,3 +1084,45 @@ Optionally, weights can be provided for the individual classifiers::
+
+The idea behind the :class:`AverageRegressor` is to combine conceptually
+different machine learning regressors and return the average predicted values.
+Such a regressor can be useful for a set of equally well performing model
```
model -> models

jnothman commented on doc/modules/ensemble.rst:
```diff
@@ -1084,3 +1084,45 @@ Optionally, weights can be provided for the individual classifiers::
+   >>> reg2 = RandomForestRegressor(random_state=1, n_estimators=10)
+   >>> reg3 = LinearRegression()
+   >>> ereg = AverageRegressor(estimators=[('gb', reg1), ('rf', reg2), ('lr', reg3)])
+   >>> reg1 = reg1.fit(X, y)
```
I don't think you should have to fit these individually.

jnothman commented on examples/ensemble/plot_average_regressor.py:
```diff
@@ -0,0 +1,44 @@
+yt = y[:20]
+plt.plot(yt, reg1.predict(xt), 'gd', yt, reg2.predict(xt), 'b^',
+         yt, reg3.predict(xt), 'ys', yt, ereg.predict(xt), 'r*')
+plt.show()
```
I think this plot needs some kind of legend indicating which is the averaged model.

jnothman commented:
We have "Voting" classifier, but "Average" regressor. Should it be "AveragingRegressor" for consistency?

jnothman commented on sklearn/ensemble/tests/test_voting.py:
```diff
@@ -98,7 +123,9 @@ def test_majority_label_iris():
 @pytest.mark.filterwarnings('ignore:The default value of n_estimators')
 def test_tie_situation():
-    """Check voting classifier selects smaller class label in tie situation."""
+    """
```
Please do not change unrelated things. It makes your contribution harder to review and may introduce merge conflicts to other pull requests.

jnothman commented on sklearn/ensemble/tests/test_voting.py:
```diff
@@ -352,6 +463,37 @@ def test_set_params():
+    ereg1 = AverageRegressor(estimators=[
+        ('lr', reg1), ('rf', reg2)], weights=[1, 2])
+
+    assert_true('lr' in ereg1.named_estimators)
```
With the adoption of pytest, we are phasing out use of test helpers `assert_equal`, `assert_true`, etc. Please use bare `assert` statements, e.g. `assert x == y`, `assert not x`, etc.

jnothman commented on sklearn/ensemble/voting.py:
```diff
@@ -1,28 +1,40 @@
-This module contains a Soft Voting/Majority Rule classifier for
+This module contains:
+ - A Soft Voting/Majority Rule classifier for
 classification estimators.
```
this should have a hanging indent

jnothman commented on sklearn/ensemble/voting.py:
```diff
@@ -339,9 +369,157 @@ def get_params(self, deep=True):
-        return np.asarray([clf.predict(X) for clf in self.estimators_]).T
+
+class AverageRegressor(BaseVoting, RegressorMixin, TransformerMixin):
+    """
```
PEP257: this line should contain a 1-line summary of what the class is, followed by a blank line.

jnothman commented on sklearn/ensemble/voting.py:
```diff
@@ -339,9 +369,157 @@ def get_params(self, deep=True):
+class AverageRegressor(BaseVoting, RegressorMixin, TransformerMixin):
+    """
+    An average regressor is an ensemble meta-estimator that fits base
+    regressors each on the whole dataset. it, then, averages the individual
```
it -> It

jnothman commented on sklearn/ensemble/voting.py:
```diff
@@ -339,9 +369,157 @@ def get_params(self, deep=True):
+        check_is_fitted(self, 'estimators_')
+        return self._predict(X)
+
+    def set_params(self, **params):
```
why does this and `get_params` need redefinition?

stsouko replied:
docstrings for voting and averaging is different.

jnothman replied:
Very few if any set_params implementations say the name of the class in their docstring.. Just change the docstring to generic

stsouko replied:
done.

Commit:
after review:

* AverageRegressor renamed to AveragingRegressor.
* user guide fixed.
* tests fixed. additional changes removed.
* docstrings fixed.
* user guide example plot fixed.

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
flake8 fixed.

Commit:
setparams getparams moved to common class.

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
whats_new fixed.

jnothman commented on sklearn/ensemble/tests/test_voting.py:
```diff
@@ -109,11 +133,26 @@ def test_tie_situation():
+    ereg = AveragingRegressor([('br', reg1), ('rf', reg2), ('gbr', reg3)])
+
+    scores = cross_val_score(ereg, X_r, y_r, cv=5, scoring='r2')
+    assert_almost_equal(scores.mean(), 0.65, decimal=2)
```
I don't know why this should be the test value. Did you calculate it by hand? If any one of the constituent estimators changes, this result and all similar ones will have to change. Rather you should test not against  some hard coded score but by checking against some simple logic for this specific set of estimators.

I also think using cross_val_score is unnecessarily complex for a test of basic functionality... And if the test failed that complexity makes it harder to debug.

jnothman replied:
Even using DummyRegressor would give a clearer test of basic functionality

stsouko replied:
changed to DummyRegressor

jnothman commented on sklearn/ensemble/tests/test_voting.py:
```diff
@@ -124,6 +163,15 @@ def test_weights_iris():
     scores = cross_val_score(eclf, X, y, cv=5, scoring='accuracy')
     assert_almost_equal(scores.mean(), 0.93, decimal=2)
```
I can see that there are similar poorly designed tests for VotingClassifier. I wish they were better.

stsouko replied:
I think VotingClassifier tests should be fixed in separate PR.

jnothman commented on sklearn/ensemble/tests/test_voting.py:
```diff
@@ -124,6 +163,15 @@ def test_weights_iris():
+    ereg = AveragingRegressor([('br', reg1), ('rf', reg2), ('gbr', reg3)],
+                              weights=[1, 2, 10])
+    scores = cross_val_score(ereg, X_r, y_r, cv=5, scoring='r2')
+    assert_almost_equal(scores.mean(), 0.67, decimal=2)
```
It's quite unusual to put tests for different estimators in the same function, unless:
* You are testing for identical behaviours between them
* You parameterize the test to account for small differences in their behaviour

You could do parameterize here if you like.

jnothman reviewed (commented):
Tests need some work

Commit:
Voting classifier tests not changed.

test_notfitted extended.
duplicates removed.
DummyRegressor used.
only for unique methods of AveragingRegressor tests added.

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
Merge remote-tracking branch 'sklearn/master'

stsouko commented:
@jnothman, tests fixed. VotingClassifier specific tests I can fix in new PR.

jnothman reviewed (commented):
Otherwise this LGTM!

jnothman reviewed (commented):
[deleted]

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
weights test fixed.
see also added.

stsouko commented:
> I think there is now no test for the unweighted case. I'm happy for both to be in the same test function, but weights=None should be tested.

> Please add a See Also entry cross-referencing between VotingClassifier and AveragingRegressor

all done.

stsouko commented:
Is this PR ready for merge? or need additional changes?

jnothman commented:
Needs an additional review.

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
Merge remote-tracking branch 'sklearn/master'

Commit:
Merge remote-tracking branch 'sklearn/master'

# Conflicts:
#	doc/whats_new/v0.21.rst

stsouko commented:
this article describe applicability of averaging regressor.
in chemoinformatics this article is one of the most significant.

https://onlinelibrary.wiley.com/doi/full/10.1002/minf.201000061

Commit:
Merge remote-tracking branch 'sklearn/master'

# Conflicts:
#	doc/whats_new/v0.21.rst
#	sklearn/ensemble/voting.py

stsouko commented:
bump.

merged

chkoar commented:
LGTM. I would name the estimator as `VotingRegressor` though.

Commit:
Merge remote-tracking branch 'sklearn/master'

# Conflicts:
#	doc/whats_new/v0.21.rst
#	sklearn/ensemble/voting.py

Commit:
doc fix.

stsouko commented:
`VotingRegressor` or `AveragingRegressor`.

I can easily change if the name is better.

jnothman commented:
I'm okay with VotingRegressor

Commit:
renamed AveragingRegressor to VotingRegressor

chkoar commented on sklearn/ensemble/voting.py:
```diff
@@ -369,21 +369,21 @@ class labels predicted by each classifier.
     .. versionadded:: 0.21
 
-    An averaging regressor is an ensemble meta-estimator that fits base
+    An voting regressor is an ensemble meta-estimator that fits base
```
A voting regressor

Commit:
docs fixes.

stsouko commented:
renaming done.

Commit:
Merge remote-tracking branch 'sklearn/master'

# Conflicts:
#	sklearn/utils/testing.py

Commit:
six removed

Commit:
tests fixed

NicolasHug commented on examples/ensemble/plot_averaging_regressor.py:
```diff
@@ -0,0 +1,51 @@
+`RandomForestRegressor`, and `LinearRegression`) and used to initialize a
+`VotingRegressor`.
+
+On plot red starred dots is averaged predictions.
```
```suggestion
 The red starred dots are the averaged predictions.
```

NicolasHug commented on examples/ensemble/plot_averaging_regressor.py:
```diff
@@ -0,0 +1,51 @@
+"""
```
Please change  file name to `plot_voting_regressor.py` instead

NicolasHug commented on examples/ensemble/plot_averaging_regressor.py:
```diff
@@ -0,0 +1,51 @@
+yt = y[:20]
+
+plt.figure()
+plt.plot(yt, reg1.predict(xt), 'gd', label='GradientBoostingRegressor')
```
What you're plotting is an error plot (y_true vs y_pred). Is that on purpose?

It would seem more natural to me to simply plot the predictions of the first 20 samples, with e.g.

```
plt.plot(reg1.predict(xt), 'gd', label='GradientBoostingRegressor') 
plt.xlabel('training samples')
```

Also the figure needs a title

NicolasHug commented on doc/whats_new/v0.21.rst:
```diff
@@ -220,6 +220,12 @@ Support for Python 3.4 and below has been officially dropped.
+- |Feature| Add :class:`ensemble.VotingRegressor`
+  which provides an equivalent of :class:`ensemble.VotingClassifier`
+  for regression problems.
+  :issue:`10868` by :user:`Ramil Nugmanov <stsouko>` and
```
```suggestion
  :issue:`12513` by :user:`Ramil Nugmanov <stsouko>` and
```

(Should be PR number)

NicolasHug commented on doc/modules/ensemble.rst:
```diff
@@ -1084,3 +1084,42 @@ Optionally, weights can be provided for the individual classifiers::
    ...                         voting='soft', weights=[2, 5, 1])
+
+
+ .. _voting_regressor:
```
```suggestion
.. _voting_regressor:
```

(Else the ref isn't valid)

stsouko replied:
old ref also works for me

NicolasHug commented on doc/modules/ensemble.rst:
```diff
@@ -1084,3 +1084,42 @@ Optionally, weights can be provided for the individual classifiers::
+Voting Regressor
+================
+
+The idea behind the :class:`VotingRegressor` is to combine conceptually
```
Link is broken, please use instead

```suggestion
The idea behind the `VotingRegressor` is to combine conceptually
```

Could you also please do the same for the first occurrence of `VotingClassifier` bove?

NicolasHug replied:
Still broken

NicolasHug commented on sklearn/ensemble/voting.py:
```diff
@@ -304,42 +364,125 @@ class labels predicted by each classifier.
+        attribute ``self.estimators_``. An estimator can be set to `None`
+        using ``set_params``.
+
+    weights : array-like, shape = [n_regressors], optional (default=`None`)
```
```suggestion
    weights : array-like, shape (n_regressors,), optional (default=`None`)
```

NicolasHug commented on sklearn/ensemble/voting.py:
```diff
@@ -304,42 +364,125 @@ class labels predicted by each classifier.
-            In addition, to setting the parameters of the ``VotingClassifier``,
-            the individual classifiers of the ``VotingClassifier`` can also be
-            set or replaced by setting them to None.
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
```
```suggestion
        X : {array-like, sparse matrix}, shape (n_samples, n_features)
```

NicolasHug commented on sklearn/ensemble/voting.py:
```diff
@@ -304,42 +364,125 @@ class labels predicted by each classifier.
-        clf2 = RandomForestClassifier()
-        eclf = VotingClassifier(estimators=[('lr', clf1), ('rf', clf2)]
-        eclf.set_params(rf=None)
+        y : array-like, shape = [n_samples]
```
```suggestion
        y : array-like, shape (n_samples,)
```

stsouko replied:
this is a new guide for documenting?
I looked in other files. there used `= []` style

NicolasHug replied:
Yes some other files still use `[]` but we're encouraging the use of parenthesis for new code (see e.g. the contirbuting guide)

NicolasHug commented on sklearn/ensemble/voting.py:
```diff
@@ -304,42 +364,125 @@ class labels predicted by each classifier.
+        y : array-like, shape = [n_samples]
+            Target values.
 
+        sample_weight : array-like, shape = [n_samples] or None
```
```suggestion
        sample_weight : array-like, shape (n_samples,) or None
```

NicolasHug commented on sklearn/ensemble/voting.py:
```diff
@@ -304,42 +364,125 @@ class labels predicted by each classifier.
-        deep : bool
-            Setting it to True gets the various classifiers and the parameters
-            of the classifiers as well
+        X : {array-like, sparse matrix} of shape = [n_samples, n_features]
```
```suggestion
        X : {array-like, sparse matrix} of shape (n_samples, n_features)
```

NicolasHug commented on sklearn/ensemble/voting.py:
```diff
@@ -304,42 +364,125 @@ class labels predicted by each classifier.
+
+        Returns
+        -------
+        y : array of shape = [n_samples]
```
```suggestion
        y : array of shape (n_samples,)
```

NicolasHug commented on sklearn/ensemble/voting.py:
```diff
@@ -304,42 +364,125 @@ class labels predicted by each classifier.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
```
```suggestion
        X : {array-like, sparse matrix}, shape (n_samples, n_features)
```

NicolasHug reviewed (commented):
First round of comments, looks good in general

NicolasHug reviewed (commented):
First round of comments, looks good in general

Commit:
Merge remote-tracking branch 'sklearn/master'

# Conflicts:
#	doc/whats_new/v0.21.rst

Commit:
fixes:

* https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r272589318
* https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r272589550
* https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r272592701
* https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r272592957
* https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r272593808

NicolasHug commented on examples/ensemble/plot_voting_regressor.py:
```diff
@@ -0,0 +1,51 @@
+plt.plot(reg3.predict(xt), 'ys', label='LinearRegression')
+plt.plot(ereg.predict(xt), 'r*', label='VotingRegressor')
+plt.ylabel('predicted')
+plt.xlabel('training samples')
```
Please remove the ticks of the x axis

NicolasHug committed:
Update sklearn/ensemble/voting.py

Co-Authored-By: stsouko <stsouko@live.ru>

NicolasHug committed:
Update sklearn/ensemble/voting.py

Co-Authored-By: stsouko <stsouko@live.ru>

NicolasHug committed:
Update sklearn/ensemble/voting.py

Co-Authored-By: stsouko <stsouko@live.ru>

NicolasHug committed:
Update sklearn/ensemble/voting.py

Co-Authored-By: stsouko <stsouko@live.ru>

NicolasHug committed:
Update sklearn/ensemble/voting.py

Co-Authored-By: stsouko <stsouko@live.ru>

Commit:
docs fixed.

Commit:
plot fixed

Commit:
ref fixed

NicolasHug commented on doc/modules/ensemble.rst:
```diff
@@ -1084,3 +1084,42 @@ Optionally, weights can be provided for the individual classifiers::
+.. figure:: ../auto_examples/ensemble/images/sphx_glr_plot_voting_regressor_001.png
+    :target: ../auto_examples/ensemble/plot_voting_regressor.html
+    :align: center
+    :scale: 75%
```
Please add an example section below, e.g.

```
.. topic:: Examples:

  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor`: an example of voting regressor.
```

NicolasHug commented on sklearn/ensemble/tests/test_voting.py:
```diff
@@ -125,6 +139,35 @@ def test_weights_iris():
+    X_r_train, X_r_test, y_r_train, y_r_test = \
+        train_test_split(X_r, y_r, test_size=.25)
+
+    reg1_y = reg1.fit(X_r_train, y_r_train).predict(X_r_test)
```
Our convention is  that `y` should refer to the ground truth values, not the predictions. Please renames these in to `reg1_pred`, `reg2_pred`, etc.

NicolasHug commented on sklearn/ensemble/tests/test_voting.py:
```diff
@@ -125,6 +139,35 @@ def test_weights_iris():
+                     weights=[1, 2, 10])
+    assert_almost_equal(ereg_y, avg, decimal=2)
+
+    ereg_n = VotingRegressor([('mean', reg1), ('median', reg2),
```
what do `n` and `o` stand for?

maybe rename into `ereg_weights_none` and `ereg_weight_equal`

NicolasHug reviewed (approved):
Last comments, LGTM otherwise

Commit:
* https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r273651011
* https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r273651723

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
https://github.com/scikit-learn/scikit-learn/pull/12513#discussion_r273647421

NicolasHug commented on doc/modules/ensemble.rst:
```diff
@@ -1084,3 +1084,46 @@ Optionally, weights can be provided for the individual classifiers::
+
+.. topic:: Examples:
+
+  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor`: an example of voting regressor.
```
```suggestion
  * :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`
```

Missing a `.py`, and after all you don't need a description since the example title will be used.

NicolasHug committed:
Update doc/modules/ensemble.rst

Co-Authored-By: stsouko <stsouko@live.ru>

jnothman commented:
Thanks @stsouko!!

NicolasHug commented:
Thanks @stsouko!

stsouko commented:
Thanks for the help in preparing PR
